File: backend/services/patient_history_service.js
const { dbRun, dbGet, dbAll } = require('../config/database');

/**
 * Ajouter un événement à l'historique patient
 * @param {string} patientIdentifier - Identifiant patient (téléphone, nom, etc.)
 * @param {number} diagnosticId - ID du diagnostic
 * @param {string} eventType - Type d'événement (diagnostic, traitement, suivi, etc.)
 * @param {Object} details - Détails de l'événement
 */
async function addPatientEvent(patientIdentifier, diagnosticId, eventType, details) {
  await dbRun(
    `INSERT INTO patient_history (patient_identifier, diagnostic_id, event_type, details)
     VALUES (?, ?, ?, ?)`,
    [patientIdentifier, diagnosticId, eventType, JSON.stringify(details)]
  );
}

/**
 * Obtenir l'historique complet d'un patient
 * @param {string} patientIdentifier - Identifiant patient
 * @returns {Promise<Array>} - Historique des événements
 */
async function getPatientHistory(patientIdentifier) {
  const history = await dbAll(
    `SELECT ph.*, d.maladie_type, d.statut, d.created_at as diagnostic_date
     FROM patient_history ph
     LEFT JOIN diagnostics d ON ph.diagnostic_id = d.id
     WHERE ph.patient_identifier = ?
     ORDER BY ph.created_at DESC`,
    [patientIdentifier]
  );
  
  return history.map(item => ({
    ...item,
    details: item.details ? JSON.parse(item.details) : null,
  }));
}

/**
 * Obtenir le profil patient consolidé
 * @param {string} patientIdentifier - Identifiant patient
 * @returns {Promise<Object>} - Profil patient
 */
async function getPatientProfile(patientIdentifier) {
  const history = await getPatientHistory(patientIdentifier);
  
  // Statistiques consolidées
  const diagnostics = history.filter(h => h.event_type === 'diagnostic');
  const totalDiagnostics = diagnostics.length;
  const positifDiagnostics = diagnostics.filter(d => d.statut === 'positif').length;
  const dernierDiagnostic = diagnostics[0];
  
  // Derniers traitements
  const traitements = history.filter(h => h.event_type === 'traitement');
  
  // Derniers suivis
  const suivis = history.filter(h => h.event_type === 'suivi');
  
  return {
    identifier: patientIdentifier,
    total_diagnostics: totalDiagnostics,
    positifs: positifDiagnostics,
    negatifs: totalDiagnostics - positifDiagnostics,
    dernier_diagnostic: dernierDiagnostic,
    derniers_traitements: traitements.slice(0, 5),
    derniers_suivis: suivis.slice(0, 5),
    historique_complet: history,
  };
}

module.exports = {
  addPatientEvent,
  getPatientHistory,
  getPatientProfile,
};

File: healther/lib/services/patient_history_service.dart
import 'dart:convert';
import 'api_service.dart';
import 'package:http/http.dart' as http;

class PatientHistoryService {
  static final PatientHistoryService _instance = PatientHistoryService._internal();
  factory PatientHistoryService() => _instance;
  PatientHistoryService._internal();

  String get baseUrl => ApiService.baseUrl;

  /// Obtenir l'historique d'un patient
  Future<List<Map<String, dynamic>>> getPatientHistory(String identifier) async {
    final response = await http.get(
      Uri.parse('$baseUrl/patient-history/$identifier'),
      headers: ApiService().headers,
    );

    if (response.statusCode != 200) {
      throw Exception('Erreur récupération historique');
    }

    final List<dynamic> data = json.decode(response.body);
    return data.map((json) => json as Map<String, dynamic>).toList();
  }

  /// Obtenir le profil patient consolidé
  Future<Map<String, dynamic>> getPatientProfile(String identifier) async {
    final response = await http.get(
      Uri.parse('$baseUrl/patient-history/$identifier/profile'),
      headers: ApiService().headers,
    );

    if (response.statusCode != 200) {
      throw Exception('Erreur récupération profil patient');
    }

    return json.decode(response.body);
  }

  /// Ajouter un événement à l'historique
  Future<void> addPatientEvent({
    required String patientIdentifier,
    int? diagnosticId,
    required String eventType,
    Map<String, dynamic>? details,
  }) async {
    final response = await http.post(
      Uri.parse('$baseUrl/patient-history'),
      headers: ApiService().headers,
      body: json.encode({
        'patient_identifier': patientIdentifier,
        'diagnostic_id': diagnosticId,
        'event_type': eventType,
        'details': details ?? {},
      }),
    );

    if (response.statusCode != 201) {
      throw Exception('Erreur ajout événement historique');
    }
  }
}
